tree/ntuple/v7/inc/ROOT/RNTupleMetrics.hxx

```diff
@@ -155,6 +158,74 @@ public:
 };
 
 
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTupleCalcFloatPerf
+\ingroup NTuple
+\brief A metric element that computes its floating point value from other counters.
+*/
+// clang-format on
+class RNTupleCalcIntPerf : public RNTuplePerfCounter {
+public:
+  using MetricFunc_t = std::function<std::pair<bool, std::int64_t>(const RNTupleMetrics &)>;
+private:
+  const MetricFunc_t fFunc;
+  std::int64_t fVal;
+
+public:
+   RNTupleCalcIntPerf(const std::string &name, const std::string &unit, const std::string &desc,
+			 MetricFunc_t &&func)
+	: RNTuplePerfCounter(name, unit, desc), fFunc(std::move(func)) { }
+
+   std::int64_t GetValueAsInt() const override { return fVal; }
+   std::string GetValueAsString() const override {
+      return std::to_string(fVal);
+   }
+
+   std::pair<bool, std::int64_t> Evaluate(const RNTupleMetrics &metrics) {
+      auto ret = fFunc(metrics);
+      if (ret.first)
+         fVal = ret.second;
+      return ret;
+   }
+};
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTupleCalcFloatPerf
+\ingroup NTuple
+\brief A metric element that computes its floating point value from other counters.
+*/
+// clang-format on
+class RNTupleCalcFloatPerf : public RNTuplePerfCounter {
+public:
+  using MetricFunc_t = std::function<std::pair<bool, double>(const RNTupleMetrics &)>;
+private:
+  const MetricFunc_t fFunc;
+  double fVal;
+
+public:
+   RNTupleCalcFloatPerf(const std::string &name, const std::string &unit, const std::string &desc,
+			 MetricFunc_t &&func)
+	: RNTuplePerfCounter(name, unit, desc), fFunc(std::move(func)) { }
+
+   std::int64_t GetValueAsInt() const override {
+      return fVal;
+   }
+   std::string GetValueAsString() const override {
+      return std::to_string(fVal);
+   }
+
+   std::pair<bool, double> Evaluate(const RNTupleMetrics &metrics) {
```

Comment:
How about making `Evaluate` private and call it as part of `GetValueAsInt()` and `GetValueAsString()`? I'm afraid that otherwise we might at some point forget to call `Evaluate` and output stale metrics.  We could use `NaN` as a value for the case that `Evaluate` returns false.

The disadvantage of making `Evaulate` part of `GetValue...` is a potential performance penalty and that we need pass the source `RNTupleMetrics` on construction and we might end up with a broken reference.